app/v2/controllers/UpscanResponseParser.scala

```diff
@@ -16,43 +16,38 @@
 
 package v2.controllers
 
-import cats.data.EitherT
 import play.api.Logging
 import play.api.libs.json.JsValue
 import play.api.mvc.BaseController
 import v2.models.errors.PresentationError
 import v2.models.responses.UpscanResponse
 
-import scala.concurrent.Future
-
 trait UpscanResponseParser {
   self: BaseController with Logging =>
 
-  def parseAndLogUpscanResponse(responseBody: JsValue): EitherT[Future, PresentationError, UpscanResponse] =
-    EitherT {
-      responseBody
-        .validate[UpscanResponse]
-        .map(
-          upscanResponse => evaluate(upscanResponse)
-        )
-        .getOrElse {
-          logger.error("Unable to parse unexpected response from Upscan")
-          Future.successful(Left(PresentationError.badRequestError("Unexpected Upscan callback response")))
-        }
-    }
+  def parseAndLogUpscanResponse(responseBody: JsValue): Either[PresentationError, UpscanResponse] =
+    responseBody
+      .validate[UpscanResponse]
+      .map(
+        upscanResponse => evaluate(upscanResponse)
+      )
+      .getOrElse {
+        logger.error("Unable to parse unexpected response from Upscan")
+        Left(PresentationError.badRequestError("Unexpected Upscan callback response"))
+      }
 
   private def evaluate(upscanResponse: UpscanResponse) =
     upscanResponse match {
       case UpscanResponse(reference, _, Some(_), Some(_), None) =>
         logger.info(s"Received a successful response from Upscan callback for the following reference: $reference")
-        Future.successful(Right(upscanResponse))
+        Right(upscanResponse)
       case UpscanResponse(reference, _, None, None, Some(failureDetails)) =>
         logger.warn(
           s"Received a failure response from Upscan callback for the following reference: $reference. Failure reason: ${failureDetails.failureReason}. Failure message: ${failureDetails.message}"
         )
-        Future.successful(Right(upscanResponse))
+        Right(upscanResponse)
```

Comment:
I am little bit confused with this case condition. so In case of Upscan response with failureDetails . should we not return 500 or Bad request or some error and should not proceed further. As looks like in case of Upscan response with failureDetails it is going to proceed further and then making call to Object store to copy the file from upscan with download url but we would not be getting download url in case of Upscan response with failureDetails? Please correct me if I misunderstood.

Reply:
If we can't parse the request from upscan, we return a bad request. 
If we can parse the request from upscan, it could be a failure or a success response. We handle that in handleUpscanSuccessResponse, where if the download URL is not present, we will be updating the trader about it when we have implemented PPNS, and returning 200. 

```
private def handleUpscanSuccessResponse(upscanResponse: UpscanResponse): EitherT[Future, PresentationError, DownloadUrl] =
     EitherT {
       Future.successful(upscanResponse.downloadUrl.toRight {
         PresentationError.badRequestError("Upscan failed to process file")
       })
     }
```

This ensures we have a downloadUrl and if we don't we return a presentation error so it exists out of the for comprehension with that. Hope that makes sense

Reply:
ok my bad. thanks.